pkg/syncer/resourcesync/controller.go

```diff
@@ -63,47 +63,35 @@ const (
 	controllerName = "kcp-syncer-resourcesync-controller"
 )
 
-type SyncerInformer struct {
-	UpstreamInformer   kcpkubernetesinformers.GenericClusterInformer
-	DownstreamInformer informers.GenericInformer
-	cancel             context.CancelFunc
-}
-
-type SyncerInformerFactory interface {
-	AddUpstreamEventHandler(handler ResourceEventHandlerPerGVR)
-	AddDownstreamEventHandler(handler ResourceEventHandlerPerGVR)
-	InformerForResource(gvr schema.GroupVersionResource) (*SyncerInformer, bool)
-	SyncableGVRs() (map[schema.GroupVersionResource]*SyncerInformer, error)
-	Start(ctx context.Context, numThreads int)
-}
-
-type ResourceEventHandlerPerGVR func(schema.GroupVersionResource) cache.ResourceEventHandler
+var _ informer.GVRSource = (*Controller)(nil)
 
 // controller is a control loop that watches synctarget. It starts/stops spec syncer and status syncer
 // per gvr based on synctarget.Status.SyncedResources.
 // All the spec/status syncer share the same downstreamNSInformer and upstreamSecretInformer. Informers
 // for gvr is started separated for each syncer.
 type Controller struct {
-	queue                        workqueue.RateLimitingInterface
-	upstreamDynamicClusterClient kcpdynamic.ClusterInterface
-	downstreamDynamicClient      dynamic.Interface
-	downstreamKubeClient         kubernetes.Interface
-
-	upstreamEventHandlers   []ResourceEventHandlerPerGVR
-	downstreamEventHandlers []ResourceEventHandlerPerGVR
-
-	syncTargetName        string
-	syncTargetClusterName logicalcluster.Name
-	syncTargetUID         types.UID
-	syncTargetLister      workloadv1alpha1listers.SyncTargetLister
-	kcpClient             clientset.Interface
-
-	syncerInformerMap map[schema.GroupVersionResource]*SyncerInformer
-	mutex             sync.RWMutex
+	queue                workqueue.RateLimitingInterface
+	downstreamKubeClient kubernetes.Interface
+
+	cachedDiscovery discovery.CachedDiscoveryInterface
+
+	syncTargetUID               types.UID
+	syncTargetLister            workloadv1alpha1listers.SyncTargetLister
+	synctargetInformerCacheSync cache.InformerSynced
+	kcpClient                   clientset.Interface
+
+	gvrs map[schema.GroupVersionResource]informer.GVRPartialMetadata
```

Comment:
gvrs of what? Are they static?

Reply:
I renamed the field. This is the list of the GVRs that should be watched by the ddsifs, and this list will feed the `GVRs()` method (implementation of the `informer.GVRSource` interface.